ports/nrf/common-hal/_bleio/bonding.c

```diff
@@ -0,0 +1,312 @@
+#define SYS_ATTR_FLAGS      (BLE_GATTS_SYS_ATTR_FLAG_SYS_SRVCS | BLE_GATTS_SYS_ATTR_FLAG_USR_SRVCS)
+
+#if BONDING_DEBUG
+void bonding_print_block(bonding_block_t *block) {
+    printf("at 0x%08lx: is_central: %1d, type: 0x%x, ediv: 0x%04x, data_length: %d\n",
+               (uint32_t) block, block->is_central, block->type, block->ediv, block->data_length);
+}
+
+void bonding_print_keys(bonding_keys_t *keys) {
+    for (size_t i = 0; i < sizeof(bonding_keys_t); i ++) {
+        printf("%x", ((uint8_t*) keys)[i]);
+    }
+    printf("\n");
+}
+#endif
+
+STATIC size_t compute_block_size(uint16_t data_length) {
+    // Round data size up to the nearest 32-bit address.
+    return sizeof(bonding_block_t) + ((data_length + 3) & ~0x3);
+}
+
+void bonding_erase_storage(void) {
+    // Erase all pages in the bonding area.
+    for(uint32_t page_address = BONDING_PAGES_START_ADDR;
+        page_address < BONDING_PAGES_END_ADDR;
+        page_address += FLASH_PAGE_SIZE) {
+        // Argument is page number, not address.
+        sd_flash_page_erase_sync(page_address / FLASH_PAGE_SIZE);
+    }
+    // Write marker words at the beginning and the end of the bonding area.
+    uint32_t flag = BONDING_FLAG;
+    sd_flash_write_sync((uint32_t *) BONDING_START_FLAG_ADDR, &flag, 1);
+    sd_flash_write_sync((uint32_t *) BONDING_END_FLAG_ADDR, &flag, 1);
+}
+
+// Given NULL to start or block address, return the address of the next valid block.
+// The last block returned is the unused block at the end.
+// Return NULL if we have run off the end of the bonding space.
+STATIC bonding_block_t *next_block(bonding_block_t *block) {
+    while (1) {
+        // Advance to next block.
+        if (block == NULL) {
+            return (bonding_block_t *) BONDING_DATA_START_ADDR;
+        } else if (block->type == BLOCK_UNUSED) {
+            // Already at last block (the unused block).
+            return NULL;
+        }
+
+        // Advance to next block.
+        block = (bonding_block_t *) ((uint8_t *) block + compute_block_size(block->data_length));
+
+        if (block >= (bonding_block_t *) BONDING_DATA_END_ADDR) {
+            // Went past end of bonding space.
+            return NULL;
+        }
+        if (block->type != BLOCK_INVALID) {
+            // Found an empty or a valid block.
+            return block;
+        }
+        // Invalid block (was erased); try again.
+    }
+}
+
+// Find the block with given is_central, type and ediv value.
+// If type == BLOCK_UNUSED, ediv is ignored and the the sole unused block at the end is returned.
+// If not found, return NULL.
+STATIC bonding_block_t *find_existing_block(bool is_central, bonding_block_type_t type, uint16_t ediv) {
+    bonding_block_t *block = NULL;
+    while (1) {
+        block = next_block(block);
+        if (block == NULL) {
+            return NULL;
+        }
+        if (block->type == BLOCK_INVALID) {
+            // Skip discarded blocks.
+            continue;
+        }
```

Comment:
I don't think you need this. next_block skips invalid blocks.

Reply:
Oh, yeah :smiley: